### ISSUE TITLE
Straight-through processing loop may cause problems

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#12957]
**Describe the bug**

Since 8.2, it is possible to deploy processes with [Undefined Task](https://docs.camunda.io/docs/next/components/modeler/bpmn/undefined-tasks/), which the workflow engine processes as a straight-through activity. When placed in a loop without a wait state (i.e. element in the process that requires the engine to wait), the workflow engine may process faster than the exporters can export the produced records.

When the workflow engine is faster than the exporters, the log grows, and Zeebe's disk space usage increases. Eventually, Zeebe Brokers may take too long to start up because the log is too large. This may lead to the unavailability of Zeebe.

We've encountered this on the Elasticsearch exporter, where ES may reject exported records. In that case, it can even result in the unavailability of Operate and Zeebe.

As a temporary workaround:
- increase the CPU/memory resources of Elasticsearch to support the higher influx of records
- increase the failure threshold Zeebe Brokers are allowed as start-up duration
- cancel the process instance with the straight-through processing loop

**To Reproduce**

> The following only works on clusters exporting to Elasticsearch with specific resource allocations. You may need to reduce the available Elasticsearch resources.

- deploy a process with a loop of undefined tasks
- create an instance of this process
- notice the number of not exported records increasing

**Expected behavior**

Process instances of processes with a straight-through processing loop should not cause problems.

**Solution proposal**

A solution has already been proposed to resolve this issue:
- #12696 

Additionally, we suggest to:
- raise an [Incident](https://docs.camunda.io/docs/next/components/concepts/incidents/) when a straight-through processing loop is detected during process execution

**Considered alternatives**

IMO, we should not reject deployments of processes without wait states because these may be valuable in Modeling/Play iteration when first designing a process.

**Comment by korthout** [comment]
Good point @Zelldon! I'll adjust the description accordingly.

**Comment by megglos** [comment]
Thanks for raising this @korthout! I wanted to follow-up on this topic this week. I would love us to have a short-term mitigation, like e.g. disabling undefined tasks until we have a proper way of at least recovering or even better detecting these situations and suspend such instances.

For short term I see these options:

1. disable undefined task processing by default on SaaS, only offer a flag to opt-in => assuming the use case is covered by  Zeebe Play (for which it is enabled)
Pro:
- no such incidents on SaaS anymore
- optional: we can make this transparent to the user by the engine rejecting processes that use undefined tasks
Con:
- the feature is effectively removed from SaaS

2. option 1 but only for all non-trial clusters => prevent an incident where a real production cluster gets impacted by such a process
Pro:
- we prevent this from happening on a production cluster
Con:
- the noise caused by trials will not disappear

3. Assess a fist iteration to detect loops of undefined tasks and reject their deployment => even if it may not cover all potential edge-cases may a first iteration be feasible that already is good-enough to prevent a majority of such processes to cause trouble?
Pro:
- may be suffice to reduce the frequency of such incidents heavily already
- error message can make this transparent to the user
- may better fit a 8.2 patch scope => prevent users from running into this issue while keeping the previous default of undefined tasks being allowed
Con:
- effort seems higher
- still a risk of an edge-case not covered causing such loop (a call-activity of a process that contains undefined tasks only 🙃 ?)

My personal favorite is option 3 and if that's too complicated option 1 asap. (I'm fine doing an out of schedule patch, for the sake of preventing this issue asap)
Option 2 is not really helping on the trial noise, meaning it is not good enough of a proper mitigation from that perspective,  and introduced feature imbalance for trial that we didn't have yet. right?

wdyt?

**Comment by megglos** [comment]
@korthout as a mitigation to this would help us to avoid repeated incidents with manual effort I would suggest to bump severity to high, wdyt?

**Comment by korthout** [comment]
@megglos The `severity` label is clearly defined. This bug does not have high severity because a workaround is available. I do see the need to prioritize it as `🚧 upcoming` (intend to work on it soon) or even as `⛔ blocker` (stop-the-world) due to the impact it has on us (incidents, alert noise, etc).

**Comment by megglos** [comment]
another option for a potential mitigation that was raised by @oleschoenburg 

4. What about having an artificial delay for undefined and manual tasks, e.g. 1s?
That would throttle such loops significantly by introducing a wait state and thus offsetting disk resource issues significantly.
From a user perspective that might be acceptable to unnoticeable. We could make it configurable to cover cases where a self-managed user cares about low latency and still wants to keep and undefined/manual task in the process
Pro:
might it be more straightforward to implement comapred to option 3
Con:
Doesn't prevent this situation but offsets any resulting resource problems

So we could decide on doing either 4 or 3 first. While 3 would prevent such situations (but not all of them) 4 would offset any resulting issues of such processes by a significant amount of time. Both together appear already good enough to mitigate the pain of such looping processes.

With these in place we may already close this and eventually shift towards https://github.com/camunda/zeebe/issues/12696 that would enable us to at least manually intervene with any rogue process.

wdyt @korthout @Zelldon ?

**Comment by oleschoenburg** [comment]
> Not sure how this should look like, since if you start to process it you need to commit all changes to it before starting on the next command

Similar to how job timeouts works for example: after the manual task is `ACTIVATED` and after some period of time has passed we write a `COMPLETE` command to the log.
That's how I thought it could work anyway, maybe that doesn't work?

**Comment by korthout** [comment]
@Zelldon is right; it wouldn't be safe to do so. Unless we add a way to write the `COMPLETE` command for any currently `ACTIVATED` undefined/manual tasks, on recovery.

**Comment by oleschoenburg** [comment]
But every `ACTIVATED` undefined/manual task should already have a `COMPLETE` follow-up command on the log. Otherwise, how would this task ever complete?

**Comment by korthout** [comment]
>But every ACTIVATED undefined/manual task should already have a COMPLETE follow-up command on the log. 

That's not correct. The proposal was to add the delay async, so when processing `ACTIVATE_ELEMENT` we append `ELEMENT_ACTIVATING` and `ELEMENT_ACTIVATED`, and also schedule a post-commit task to append `COMPLETE_ELEMENT` after a delay (this may be lost).

**Comment by oleschoenburg** [comment]
Ah sorry, I misunderstood and thought you were talking about the upgrade from old to new behavior.

In general I agree, this would require persistence of activated manual/undefined tasks so that they can be completed eventually. Again, similar to job timeouts.

**Comment by Zelldon** [comment]
So to summarize the proposal (just that I get it right):

 * There will be another checker/consumer which is scheduled via the ProcessingScheduleService (in order to be decoupled of the processing). Lets call it U-COMPLETER
 * The processing of undefined tasks is split up, it will only produce ACTIVATE_ACTIVATED on processing
 * During activation we need to store the record and all necessary data in a new column family (for the U-COMPLETER)
 * The U-COMPLETER will write based on the data in the column family the complete commands, in order to continue the processing of the instance.

**Comment by korthout** [comment]
ZPA triage:
- let's split up this issue into the different solution proposals
- we like the proposal to reject deployments of processes with an undefined task loop but there are simpler solutions, so we'll focus on those first
- we like the proposal to slow down undefined task, but it's unclear to us whether we should do this sync or async.
  - sync: this blocks the stream processor a bit but is a simpler solution
  - async: adds additional complexity over sync because we have to care about the unreliability of scheduled post-commit tasks
  - @camunda/zeebe-distributed-platform, do you have an opinion about the performance loss?
- we should check what impact our solutions have on Zeebe Play
- we will focus on the quick fix in the upcoming iteration

**Comment by megglos** [comment]
> sync: this blocks the stream processor a bit but is a simpler solution
> async: adds additional complexity over sync because we have to care about the unreliability of scheduled post-commit tasks

sync would not be an option due to the impact on overall processing latency

**Comment by megglos** [comment]
I'm also in all in for rolling out a 8.2.X patch & an 8.2.0-alpha2.1 (did we ever do an out of cadence alpha before?) as soon as a mitigation is available

**Comment by korthout** [comment]
>sync would not be an option due to the impact on overall processing latency/throughput

Is Undefined Task really an element used in production where performance (latency/throughput matters?

**Comment by remcowesterhoud** [comment]
I've created https://github.com/camunda/zeebe/issues/12993 as a separate issue to delay undefined tasks.

**Comment by korthout** [comment]
Thanks @remcowesterhoud, marking this issue as `later` priority, as #12993 is split out with higher prio.

**Comment by oleschoenburg** [comment]
> I imagine Undefined Task only has use in Zeebe Play, or as a lonely element in a production process where a small latency (~10-50ms) doesn't really matter.

I disagree that it's fine for production workloads. If a user "accidentally" deploys and starts such as process that loops on an undefined task, it could seriously impact the performance of an entire partition because the partition will repeatedly sleep. Even worse with batch processing enabled where it's not just one sleep but many (like 10 or 100).

**Comment by megglos** [comment]
ZDP-Triage:
- moving to backlog now as @remcowesterhoud is looking into this

**Comment by remcowesterhoud** [comment]
As the quick-fix is done and getting released as I type this, I'm going to remove my assignment from this issue. We'll still need to make a proper solution, so I will move this back to the teams inbox to triage it again.

**Comment by Zelldon** [comment]
Great work @remcowesterhoud

**Comment by korthout** [comment]
ZPA triage:
- with a quick fix available, we think other issues have higher priority than this
- we'll increase priority if this does occur again
- marking as `later`

**Comment by megglos** [comment]
ZDP-Triage:
- loop detection was a first band-aid and we still lack measures to handle situations not covered by validation
- measures from the steam platform might be possible, e.g. detect high processing load on the same key, it needs to be discussed if this is the right direction though
- most solutions likely require collab between both teams

**Comment by megglos** [comment]
https://github.com/camunda/zeebe/issues/12696 is related to this as well, will put focus on that one first

**Comment by megglos** [comment]
ZDP-Planning:
- is related to #12560 and would be mitigated if processing is rate limited
- we have alerts on noop loops => we can cancel such instances proactively

**Comment by megglos** [comment]
given it's mitigated by validation and having the ability to forcefully terminate such instances with https://github.com/camunda/zeebe/issues/12696 

further we plan to invest into https://github.com/camunda/zeebe/issues/12560 first

moving to blocked
